src/cat/cat.rs

```diff
@@ -48,280 +45,202 @@ pub fn uumain(args: Vec<String>) -> i32 {
     } else {
         NumberingMode::NumberNone
     };
-    let show_nonprint = matches.opts_present(&["A".to_owned(), "e".to_owned(),
-                                              "t".to_owned(), "v".to_owned()]);
-    let show_ends = matches.opts_present(&["E".to_owned(), "A".to_owned(),
-                                          "e".to_owned()]);
-    let show_tabs = matches.opts_present(&["A".to_owned(), "T".to_owned(),
-                                          "t".to_owned()]);
+    let show_nonprint =
+        matches.opts_present(&["A".to_owned(), "e".to_owned(), "t".to_owned(), "v".to_owned()]);
+    let show_ends = matches.opts_present(&["E".to_owned(), "A".to_owned(), "e".to_owned()]);
+    let show_tabs = matches.opts_present(&["A".to_owned(), "T".to_owned(), "t".to_owned()]);
     let squeeze_blank = matches.opt_present("s");
     let mut files = matches.free;
     if files.is_empty() {
         files.push("-".to_owned());
     }
 
-    exec(files, number_mode, show_nonprint, show_ends, show_tabs, squeeze_blank);
+    if show_tabs || show_nonprint || show_ends || squeeze_blank ||
+       number_mode != NumberingMode::NumberNone {
+        write_lines(files,
+                    number_mode,
+                    squeeze_blank,
+                    show_ends,
+                    show_tabs,
+                    show_nonprint);
+    } else {
+        write_fast(files);
+    }
+    pipe_flush!();
 
     0
 }
 
-#[derive(Eq, PartialEq)]
+#[derive(PartialEq)]
 enum NumberingMode {
     NumberNone,
     NumberNonEmpty,
     NumberAll,
 }
 
-fn write_lines(files: Vec<String>, number: NumberingMode, squeeze_blank: bool,
-               show_ends: bool) {
+fn open(path: &str) -> Option<(Box<Read>, bool)> {
+    if path == "-" {
+        let stdin = stdin();
+        return Some((Box::new(stdin) as Box<Read>, is_stdin_interactive()));
+    }
 
-    let mut line_counter: usize = 1;
+    match File::open(path) {
+        Ok(f) => Some((Box::new(f) as Box<Read>, false)),
+        Err(e) => {
+            (writeln!(&mut stderr(), "cat: {0}: {1}", path, e.to_string())).unwrap();
+            None
+        }
+    }
+}
 
-    for (mut reader, interactive) in files.iter().filter_map(|p| open(&p[..])) {
+fn write_fast(files: Vec<String>) {
+    let mut writer = stdout();
+    let mut in_buf = [0; 1024 * 64];
 
-        let mut in_buf  = [0; 1024 * 31];
-        let mut out_buf = [0; 1024 * 64];
-        let mut writer = UnsafeWriter::new(&mut out_buf[..], stdout());
-        let mut at_line_start = true;
-        let mut one_blank_kept = false;
+    for (mut reader, _) in files.iter().filter_map(|p| open(&p[..])) {
```

Comment:
You are not right here for two reasons:

1. `files.iter().filter_map(|p| open(&p[..]))` is no more than `struct` which implement `Map`. On each step of the loop the `next` method is called.
2. `reader` variable is local to loop body, so `reader.drop()` is used at the end of every iteration. As far as I know, `drop` method automatically closes the file.